### ISSUE TITLE
`compute_gradient` Problems

### ISSUE THREAD

**Issue by JonathonMisiewicz** [psi4/psi4#2223]
With apologies for the vague title, there are some issues with the `compute_gradient` function. They're connected, so it's best to have a single topic for these:

1. Not all modules that support gradients support the `compute_gradient` API. `dct` and Rob's `dfmp2` do, but most other modules don't. We learned in #2200 that RHF wavefunctions don't, though `SCFDeriv` and `Dispersion` "wavefunctions" seem to. (See `scfgrad` and `libdisp`, respectively.)
2. The `compute_gradient` calls coming out of `dct` and `dfmp2` don't even work without crashing! Minimal not-working example:
```
molecule {
O
H 1 1.0 
H 1 1.0 2 104.5
}

wfn = energy('mp2/sto-6g', return_wfn=True)[1] # Can switch "mp2" to "dct" and get a similar crash
wfn.compute_gradient()
```
3. Computing a derivative _reliably_ requires assumptions about what intermediates are available and what the Hamiltonian is. For example, we need to know that the file of "T" amplitudes is the same as that optimized during the energy computation, and that the energy was computed without relativistic corrections. We currently store whether the relativistic global is _currently_ enabled, but not whether relatvistic terms were added during the compute step.

I'm open to alternatives, but I favor taking `compute_gradient` out of the API entirely. It's not consistently in the API, it doesn't get simple cases right when it's there, and the assumptions required by such a code don't make it a good candidate for being in the API in the first place.

**Comment by loriab** [comment]
I agree it doesn't belong in the user's API. Seems reasonably useful for programmer's API. Trouble is we don't have much of a differentiation, other than warning in the exported docstring. Were you thinking remove it from `Wavefunction` and only have it in occasional subclasses?

**Comment by JonathonMisiewicz** [comment]
As best as I can tell, it's not actually _used_ Python-side, except for `dfmp2`. My first choice would be to keep the function C-side, but not export it Py-side. If that would be a problem for plugins, sticking a warning in the `docstring` is the best solution. I can't imagine how we would have separate user API and programmer API.

The only plugins I imagine using `compute_gradient` are V2RDM and Forte. V2RDM doesn't call it Py-side. Forte does, but I can personally take responsibility for moving that to C-side, if Francesco approves.

**Comment by PeterKraus** [comment]
Is there a case when a user (without deeper knowledge of the Psi4 codebase) would want to use `compute_gradient` instead of a plain-old `gradient` call? Unless there's a good use case, I'm for simplifying the API. 

If it's a matter of "you can save time by reusing intermediates" with `compute_gradient`, perhaps a more systematic way would be to shove those intermediates onto the `wfn` and then feed that into `gradient`, no?

**Comment by JonathonMisiewicz** [comment]
Not that I'm aware of. There are three reasons somebody might be tempted to use a manual `compute_gradient` call.
1. _Method developer reason_. You are a plugin developer and don't know you can edit `proc_table` to tell `gradient` how to run for your new method. Not a good reason because you actually can.
2. _Method developer reason_. You are writing a gradient method and want your gradient function to enter the C-code to do your energy, return to the driver, then re-enter the C-code to do your gradient. This is the best reason I've come up with, and while I'm sympathetic, I just don't think it's worth the danger of exposing this function to completely general Python use. The gradient machinery should only ever be called after your energy computation finishes.
3. _User reason_. You want to run some additional commands between your energy and your gradient computation. I just can't imagine a use case for this _other than_ bypassing some sanity check. There isn't a case where separating the calls saves times. Even if there were, shoving intermediates onto a `wfn` only works for SCF (not sure about general DFT functionals). Even MP2 gradient theory needs intermediates too large to store on the `wfn`. The best I've come up with is that you're doing SCF, serialize the wavefunction for later analysis, and then decide you want a gradient after that later analysis. But even in that case, the serialized wavefunction has the converged orbitals, so just use `gradient`. I would expect minor computational overhead for the energy step. (I suppose there is the cost of re-building the TEI?)

Since this would affect Forte and heads off a possible bug in relativistic computations, it's probably worth asking @fevangelista for input on this.

**Comment by loriab** [comment]
`compute_gradient` is also in use py-side for dispersion. perhaps we could cease to export the fn for general wfn and only keep export for libdisp and dfmp2. that reduces the risk of relativistic misuse to one method.

**Comment by JonathonMisiewicz** [comment]
That would be an improvement? I'd need to look at `libdisp` to see if we can remove it more aggressively, but I know for sure that taking it out of `dfmp2` is a fairly simple refactor.
